### PR TITLE
Implement directed edge swap

### PR DESCRIPTION
**Pull request by ben-heil**
I'm working on a project where I'll be swapping edges to build a baseline distribution of graphs. However, the graphs are directed, so [double_edge_swap](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.swap.double_edge_swap.html) doesn't work. I found an old issue ([#490](https://github.com/networkx/networkx/issues/490)) that mentions a three-edge swapping solution from [this paper](https://arxiv.org/pdf/0905.4913.pdf). This PR implements that algorithm, documents it, and adds tests.

A few questions:
- I named the new method `directed_edge_swap` to differentiate it from the other two swapping functions. Is that a reasonable name, or should it be `triple_edge_swap`?
- Is there a recommended method/citation style for adding references to the documentation? 
- The new method avoids creating parallel edges to allow it to work with DiGraph objects, but this leads to situations where it won't perform a swap that would be valid in a MultiDiGraph. I did it that way to be consistent with `double_edge_swap`, but wanted to make sure that that is the intended behavior since neither `double_edge_swap` nor the new `directed_edge_swap` prohibit Multi*[Di]Graph objects. 
- When I edit `release_dev.rst`, should this be listed as an improvement or an API change?

### CODE REVIEW

**Comment by dschult** [comment]
Thanks for this!  Merging your code to implement the algorithm pointed to in an old issue is sooo much better than having a link in an old issue.  This will help many people I suspect.

- I like the name `directed_edge_swap` rather than `triple_edge_swap`.
- The formatting of the citation is not standardized, but should include all the information needed to find the main published page for the article. It'd great to have a link that is permanent (like the doi) too.
- list this as an improvement in the release_dev file because it isn't changing an existing function.
- the handling of multigraph is something I will have to review to restore my memory of how that is handled.  Hopefully soon...

File: networkx/algorithms/swap.py [hunk 1]
```diff
@@ -6,7 +6,121 @@
...
+
+    seed : integer, random_state, or None (default)
+        Indicator of random number generation state.
+        See :ref:`Randomness<randomness>`.
+
+    Returns
+    -------
+    G : DiGraph
+       The graph after triple-edge swaps.
+
```
Comment (rossbar) [inline networkx/algorithms/swap.py:42]: It'd be good to add a `Raises` section here with all of the conditions that result in exceptions, e.g.

```rst
Raises
------
NetworkXError
    If `G` is not directed, or
    If `nswap` > `max_tries`, or
    If there are fewer than 4 nodes in `G`
```

**Review by rossbar — COMMENTED** [review]
Just a couple thoughts from a quick readthrough. Seems like a nice new feature!

File: networkx/algorithms/swap.py [hunk 2]
```diff
@@ -6,7 +6,121 @@
 
 import networkx as nx
 
-__all__ = ["double_edge_swap", "connected_double_edge_swap"]
+__all__ = ["double_edge_swap", "connected_double_edge_swap", "directed_edge_swap"]
+
+
+@py_random_state(3)
+def directed_edge_swap(G, nswap=1, max_tries=100, seed=None):
```
[the remark this answers is not in the document]
Comment (ben-heil) [reply networkx/algorithms/swap.py:13]: Sounds good!

File: networkx/algorithms/swap.py [hunk 3]
```diff
@@ -6,7 +6,130 @@
...
+       Maximum number of attempts to swap edges
+
+    seed : integer, random_state, or None (default)
+        Indicator of random number generation state.
+        See :ref:`Randomness<randomness>`.
+
+    Returns
+    -------
+    G : DiGraph
+       The graph after triple-edge swaps.
```
Comment (dschult) [inline networkx/algorithms/swap.py:41]: The term triple-edge swaps hasn’t been used before in this doc_string. So you should either add a sentence above saying that directed edge swaps are also known as triple edge swaps, or change this to directed edge swaps (are probably both of these changes).

File: networkx/algorithms/swap.py [hunk 4]
```diff
@@ -6,7 +6,130 @@
...
+
+    References
+    ----------
+    .. [1] Erdős, Péter L., et al. “A Simple Havel-Hakimi Type Algorithm to Realize
+           Graphical Degree Sequences of Directed Graphs.” ArXiv:0905.4913 [Math],
+           Jan. 2010. https://doi.org/10.48550/arXiv.0905.4913.
+    .. [2] “Combinatorics - Reaching All Possible Simple Directed Graphs with a given
+           Degree Sequence with 2-Edge Swaps.” Mathematics Stack Exchange,
+           https://math.stackexchange.com/questions/22272/. Accessed 30 May 2022.
+    """
```
Comment (dschult) [inline networkx/algorithms/swap.py:66]: You should use the `@nx.utils.not_implemented_for(“undirected”)` decorator to check for the graph type of the incoming graph.   This should have been done for `double_edge_swap` too. But fixing that can be another PR.

File: networkx/algorithms/swap.py [hunk 5]
```diff
@@ -6,7 +6,130 @@
...
+        )
+    if nswap > max_tries:
+        raise nx.NetworkXError("Number of swaps > number of tries allowed.")
+    if len(G) < 4:
+        raise nx.NetworkXError("Graph has less than four nodes.")
+
+    # Instead of choosing uniformly at random from a generated edge list,
+    # this algorithm chooses nonuniformly from the set of nodes with
+    # probability weighted by degree.
+    n = 0
```
Comment (dschult) [inline networkx/algorithms/swap.py:80]: Can we name `nx something more descriptive.  Perhaps `tries` since we use `max_tries`.

File: networkx/algorithms/swap.py [hunk 6]
```diff
@@ -6,7 +6,130 @@
...
+            G.add_edge(start, third)
+            G.add_edge(third, second)
+            G.add_edge(second, fourth)
+            G.remove_edge(start, second)
+            G.remove_edge(second, third)
+            G.remove_edge(third, fourth)
+            swapcount += 1
+
+        if n >= max_tries:
+            e = (
```
Comment (dschult) [inline networkx/algorithms/swap.py:126]: Can we name `e` something else since we typically save e for an exception rather than a message?
We often use `msg.  But really, you can write the exception with it’s message all on one line and let `black` choose how to format it.

File: networkx/algorithms/swap.py [hunk 7]
```diff
@@ -6,7 +6,130 @@
...
+                f"Maximum number of swap attempts ({n}) exceeded "
+                f"before desired swaps achieved ({nswap})."
+            )
+            raise nx.NetworkXAlgorithmError(e)
+        n += 1
```
Comment (dschult) [inline networkx/algorithms/swap.py:131]: You have `n += 1` In many places in this loop. Can you do it as the very first line in the loop? You could also move the check for `max_tries` up there too if you want them to be near each other.

File: networkx/algorithms/swap.py [hunk 8]
```diff
@@ -6,7 +6,130 @@
...
+    while swapcount < nswap:
+        # choose source node index from discrete distribution
+        start_index = discrete_sequence(1, cdistribution=cdf, seed=seed)[0]
+        start = keys[start_index]
+
+        # If the given node doesn't have any out edges, then there isn't anything to swap
+        if G.out_degree(start) == 0:
+            n += 1
+            continue
+        _, second = seed.choice(list(G.out_edges(start)))
```
Comment (dschult) [inline networkx/algorithms/swap.py:95]: This idiom uses `out_edges to find successors. It would be better to use the `list(G.succ[start])` idiom.  Or maybe better, since this lookup could be done quite often, a preprocessing step could build a dict of successor lists. That would avoid constructing the same list many times:  `nbrs = {n: list(G.succ[n]) for n in G}` and then 1st second = seed.choice(nbrs[start])`.  Similarly for the other choice lines.

File: networkx/algorithms/swap.py [hunk 9]
```diff
@@ -6,7 +6,130 @@
...
+        if G.out_degree(third) == 0:
+            n += 1
+            continue
+        _, fourth = seed.choice(list(G.out_edges(third)))
+
+        if start == second or second == third or third == fourth:
+            n += 1
+            continue
+
+        if (
```
Comment (dschult) [inline networkx/algorithms/swap.py:111]: Similarly here, use `G.succ` rather than `out_edges to check for successors.  This time you don’t want a list of the successors, a dict is better. So use `G.succ` Rather than this dict-of-lists `nbrs
```python
If third not in G.succ[start] and fourth not in G.succ[second] and second not in G.succ[third]:
```
You can check the timing of whether it is significant to create a variable to hold `G.succ`.  Sometimes it is better to define at the top of the function `G_succ = G.succ` to avoid the same lookup over and over.

**Review by dschult — COMMENTED** [review]
I went through the code parts of the new function and put some comments below. Feel free to push back on any of them. It looks like some of my comments should probably be applied to other functions in the same module, but those can be done in a different PR.

I also looked at the tests and I don’t have suggestions there!  Nice!  I would prefer if this whole test module avoided sorting based on nodes since there is no reason to sort them (just use `all()` on a list of equality tests) but again, that is something for another PR if at all…

Thanks!

File: networkx/algorithms/swap.py (hunk 8 above)
Comment (ben-heil) [reply networkx/algorithms/swap.py:95]: I updated to use the `G.succ` idiom, but avoided prebuilding the lists for this and the following comment. 

My first thought was that in situations with many nodes and few swaps, prebuilding the list would be slow and a cache would be preferred. I started to implement a cache, but realized that most of the time when we get a cache hit we'll immediately have to rebuild the list for that cache entry anyway since we're modifying that node's successors. 

I think for sparse graphs or graphs with a lot of isolated cycles it might still be worth implementing a cache, but for graphs where most iterations result in a successful swap it's a large amount of memory overhead exchanged for little to no runtime speedup

File: networkx/algorithms/swap.py (hunk 9 above)
Comment (ben-heil) [reply networkx/algorithms/swap.py:111]: Looks like `G_succ = G.succ` didn't make much of a difference here. The original took ~262 seconds, and the  `G_succ = G.succ` version took ~267. 
``` python
import copy
import networkx as nx
import time

graph = nx.gnp_random_graph(2000, 0.5, directed=True)
graph2 = copy.deepcopy(graph)

start = time.time()
G = nx.directed_edge_swap(graph, nswap=1000000, max_tries=1e16, seed=1)
end = time.time()
print(end - start)
start = time.time()
G = nx.other_edge_swap(graph2, nswap=1000000, max_tries=1e16, seed=1)
end = time.time()
print(end - start)

```

**Comment by ben-heil** [comment]
Sorry for the slow response! I went out of town but I'm back now.

File: networkx/algorithms/swap.py (hunk 8 above)
Comment (dschult) [reply networkx/algorithms/swap.py]: You are right that caching or (prebuilding) the lists will not be effective here because the graph is being morphed so the overhead of updating the lists would likely be more than simply rebuilding it when needed. :}

File: networkx/algorithms/swap.py [hunk 10]
```diff
@@ -6,7 +6,120 @@
...
+
+    Parameters
+    ----------
+    G : DiGraph
+       A directed graph
+
+    nswap : integer (optional, default=1)
+       Number of three-edge (directed) swaps to perform
+
+    max_tries : integer (optional)
```
Comment (dschult) [inline networkx/algorithms/swap.py:32]:
```suggestion
    max_tries : integer (optional, default=100)
```

File: networkx/algorithms/tests/test_swap.py [hunk 11]
```diff
@@ -5,6 +5,24 @@
 # random.seed(0)
 
 
+def test_directed_edge_swap():
+    graph = nx.gnp_random_graph(200, 0.5, directed=True)
+    in_degrees = sorted((n, d) for n, d in graph.in_degree())
+    out_degrees = sorted((n, d) for n, d in graph.out_degree())
+    G = nx.directed_edge_swap(graph, nswap=40, max_tries=500)
```
Comment (dschult) [inline networkx/algorithms/tests/test_swap.py:12]: Any tests that involve random numbers can be very difficult to track if they start failing. The test fails, and its hard to know why and then rerunning the test passes it again. So our policy has been to use a seed for the tests we are going to run regularly. This will be needed for both graph creation (`gnp_random_graph`) and for the swap. I guess this also means we can eliminate the test which explicitly tests that `seed` works. :}

File: networkx/algorithms/tests/test_swap.py [hunk 12]
```diff
@@ -53,6 +71,31 @@ def test_connected_double_edge_swap_star_low_window_threshold():
...
+
+
+def test_directed_edge_swap_tries():
+    with pytest.raises(nx.NetworkXError):
+        G = nx.directed_edge_swap(
+            nx.gnp_random_graph(3, 0.5, directed=True), nswap=1, max_tries=0
+        )
+
+
+def test_directed_edge_directed():
```
Comment (dschult) [inline networkx/algorithms/tests/test_swap.py:86]:
```suggestion
def test_directed_exception_undirected():
```

File: networkx/algorithms/tests/test_swap.py [hunk 13]
```diff
@@ -53,6 +71,31 @@ def test_connected_double_edge_swap_star_low_window_threshold():
...
 
+def test_directed_edge_swap_small():
+    with pytest.raises(nx.NetworkXError):
+        G = nx.directed_edge_swap(nx.gnp_random_graph(3, 0.5, directed=True))
+
+
+def test_directed_edge_swap_tries():
+    with pytest.raises(nx.NetworkXError):
+        G = nx.directed_edge_swap(
+            nx.gnp_random_graph(3, 0.5, directed=True), nswap=1, max_tries=0
```
Comment (dschult) [inline networkx/algorithms/tests/test_swap.py:82]: Just to avoid the randomness issue. If there's a reason for the randomness leave it as `gnp_random_graph`
```suggestion
            nx.path_graph(3, create_using=nx.DiGraph), nswap=1, max_tries=0
```

File: networkx/algorithms/tests/test_swap.py [hunk 14]
```diff
@@ -53,6 +71,31 @@ def test_connected_double_edge_swap_star_low_window_threshold():
     assert degrees == sorted(d for n, d in graph.degree())
 
 
+def test_directed_edge_swap_small():
+    with pytest.raises(nx.NetworkXError):
+        G = nx.directed_edge_swap(nx.gnp_random_graph(3, 0.5, directed=True))
```
Comment (dschult) [inline networkx/algorithms/tests/test_swap.py:76]: Perhaps better to use a known graph than a random graph. It also makes it easier to see what the function is testing.
```suggestion
        G = nx.directed_edge_swap(nx.path_graph(3, create_using=nx.DiGraph)
```

File: networkx/algorithms/swap.py [hunk 15]
```diff
@@ -6,7 +6,120 @@
...
+    -----
+    Does not enforce any connectivity constraints.
+
+    The graph G is modified in place.
+
+    References
+    ----------
+    .. [1] Erdős, Péter L., et al. “A Simple Havel-Hakimi Type Algorithm to Realize
+           Graphical Degree Sequences of Directed Graphs.” ArXiv:0905.4913 [Math],
+           Jan. 2010. https://doi.org/10.48550/arXiv.0905.4913.
```
Comment (dschult) [inline networkx/algorithms/swap.py:63]:
```suggestion
           Jan. 2010. https://doi.org/10.48550/arXiv.0905.4913.
           Published  2010 in Elec. J. Combinatorics (17(1)). R66.
           http://www.combinatorics.org/Volume_17/PDF/v17i1r66.pdf
```

**Review by dschult — COMMENTED** [review]
This looks good!  Thanks!   
I have some nits and suggestions from the tests and docstring.
Then I think it is ready for me to approve and we'll get someone else to look at it too.

File: networkx/algorithms/tests/test_swap.py (hunk 11 above)
Comment (ben-heil) [reply networkx/algorithms/tests/test_swap.py:12]: Makes sense! I went ahead and converted `test_directed_edge_swap` to use a path graph, added a seed to its `directed_edge_swap` call, and removed the seed test

**Review by dschult — COMMENTED** [review]
One more thing -- (famous last words)...
You should add this function to the documentation indexing system at:
`doc/reference/algorithms/swap.rst` so that it gets pulled into the main documentation.

I suppose you could check some equality of nodes before the others are constructed.  That is, `if start == second: continue` could go just after `second` is chosen, and similar for `second == third`. It probably doesn't speed it up much, but who knows...

I think I'm fine with the PR after these.
Thanks!

**Review by dschult — APPROVED** [review]
Looks good to me!

**Comment by rossbar** [comment]
I'm too slow on the draw :) - I'll open that issue, thanks again @ben-heil!
